Added `is_fragment` parameter to the `LexborHTMLParser` constructor, allowing users to specify whether to parse input as a full HTML document or as a fragment!

pygarap opened the pull request:
This pull request introduces significant enhancements to the `LexborHTMLParser` API, focusing on improved HTML fragment parsing, expanded type stubs, and better documentation. The changes add support for parsing HTML fragments (not just full documents), clarify and expand the Python interface, and improve the maintainability of the codebase.

**HTML Fragment Parsing Support:**

* Added an `is_fragment` parameter to the `LexborHTMLParser` constructor, allowing users to specify whether the input should be parsed as a full HTML document or as an HTML fragment. When parsing as a fragment, required elements like `<html>`, `<head>`, and `<body>` are not automatically inserted. [[1]](diffhunk://#diff-39fbdc153983fde3ef86a8139f2e47f7da0d461f2f2cec33233ba4b04a51519fL29-R103) [[2]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL802-R906)
* Introduced new internal methods (`_new_html_document`, `_parse_html_document`, and `_parse_html_fragment`) to handle the creation and parsing of both full documents and fragments. [[1]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR264-R269) [[2]](diffhunk://#diff-39fbdc153983fde3ef86a8139f2e47f7da0d461f2f2cec33233ba4b04a51519fL29-R103)
* Updated Cython and C header definitions to expose fragment parsing functions from Lexbor. [[1]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR222-R229) [[2]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR318)

**Python Type Stubs and Documentation Improvements:**

* Expanded and clarified the `selectolax/lexbor.pyi` type stubs for `LexborHTMLParser`, including detailed docstrings for all public methods and properties, improved parameter types, and return value annotations. [[1]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1-R2) [[2]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL802-R906) [[3]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR917-R936) [[4]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL863) [[5]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL885-R971) [[6]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL907-R993) [[7]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL929-R1015) [[8]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL948-R1034) [[9]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1045-R1053) [[10]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL974-L1005) [[11]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1150-R1167) [[12]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1189-R1223)
* Added new properties and methods to the type stubs, such as `inner_html` (with getter and setter), and improved the documentation and signatures for existing methods like `text`, `tags`, `strip_tags`, `merge_text_nodes`, and others.
* Improved the accuracy of default values and parameter types (e.g., `default` is now `Any` instead of `bool` in several methods). [[1]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL885-R971) [[2]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL907-R993) [[3]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL929-R1015)
* Updated docstrings to clarify return types, error conditions, and usage examples. [[1]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL802-R906) [[2]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR917-R936) [[3]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL948-R1034) [[4]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1045-R1053) [[5]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL974-L1005) [[6]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1150-R1167) [[7]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1189-R1223)

**Code Quality and Linting:**

* Added `cython-lint` to the `Makefile` lint target to ensure Cython code style consistency.

These changes collectively make the parser more flexible, easier to use, and better documented for both full HTML documents and fragments.

**References:**  
[[1]](diffhunk://#diff-39fbdc153983fde3ef86a8139f2e47f7da0d461f2f2cec33233ba4b04a51519fL29-R103) [[2]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL802-R906) [[3]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR264-R269) [[4]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR222-R229) [[5]](diffhunk://#diff-2e3fd45bdd972459a4bf6b6e93af90c506b4d08498b6ad0faab92b54e1a3181fR318) [[6]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1-R2) [[7]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR917-R936) [[8]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL863) [[9]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL885-R971) [[10]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL907-R993) [[11]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL929-R1015) [[12]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL948-R1034) [[13]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1045-R1053) [[14]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eL974-L1005) [[15]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1150-R1167) [[16]](diffhunk://#diff-7cc3778e34957e3ada45ae26778d3514177f1dec886ef4d8c223812380e3493eR1189-R1223) [[17]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R52)

rushter commented on selectolax/lexbor.pxd:
```diff
@@ -111,32 +111,32 @@ cdef extern from "lexbor/html/html.h" nogil:
 
     ctypedef enum lxb_dom_node_type_t:
-        LXB_DOM_NODE_TYPE_ELEMENT                = 0x01
-        LXB_DOM_NODE_TYPE_ATTRIBUTE              = 0x02
```
Such formatting is intentional. Please avoid changing formatting everywhere. It makes it harder to review and keeps unrelated changes in the same commits.

pygarap commented:
> I will check this later, but please avoid adding new features/parameters without creating an issue first. We don't want to introduce a ton of new APIs. That makes it harder to support in the future and confuses users when there are multiple options available.

@rushter Thank you for the feedback!

To clarify, `with_top_level_tags` is enabled by default, but with the default value, the behavior stays exactly the same as before. The behavior only changes if a user explicitly sets `with_top_level_tags=False`, so this is an entirely optional switch.
My goal with this parameter is to address:

1. [https://github.com/django-components/django-components/pull/823#issue-2723448435](https://github.com/django-components/django-components/pull/823#issue-2723448435)
2. [https://github.com/rushter/selectolax/discussions/160#discussion-8220756](https://github.com/rushter/selectolax/discussions/160#discussion-8220756)

If you prefer, I can open an issue for this and update the pull request to reference it.

pygarap commented:
> @pygarap Let's rename the argument to `is_fragment` and flip the boolean. `with_top_level_tags` is too confusing.
> 
> And we need to document it to something like this:
> 
> ```
> 
> When ``True``, treats input as an HTML fragment. Does not add required HTML5 tags when they are missing.
> 
> When set to `False`, it parses HTML fragments as well, but differently. 
> Adds missing required tags (such as '<html>', '<head>', '<body>') to the tree according to the HTML5 specification.
> Same as in browsers when they render HTML fragments as pages.
> ```

@rushter Thank you for the suggestions.
I have renamed the argument to is_fragment, flipped the boolean, and updated the documentation to match your wording. I have pushed the changes.
Please let me know if you would like me to change or clarify anything else.

rushter commented on selectolax/lexbor.pyx:
```diff
@@ -336,15 +572,15 @@ cdef class LexborHTMLParser:
     def clone(self):
-        """Clone the current node.
-
-        You can use to do temporary modifications without affecting the original HTML tree.
```
Please keep those comments.

pygarap replied:
@rushter Sorry about removing them earlier, and thanks for pointing that out.

I have restored those comments as requested and pushed the updated changes.
Please let me know if you would like me to adjust anything else.
